books/RayTracingTheNextWeek.html

```diff
@@ -2486,36 +2473,57 @@
 
 <div class='together'>
 Quads are flat, so their axis-aligned bounding box will have zero thickness in one dimension if the
-quad lies in the XY, YZ, or ZX plane. This can lead to numerical problems with ray intersection, but
+quad lies in the XY, YZ, or XZ plane. This can lead to numerical problems with ray intersection, but
 we can address this by padding any zero-sized dimensions of the bounding box. Padding is fine
 because we aren't changing the intersection of the quad; we're only expanding its bounding box to
 remove the possibility of numerical problems, and the bounds are just a rough approximation to the
-actual shape anyway. To this end, we add a new `aabb::pad()` method that remedies this situation:
+actual shape anyway. To remedy this situation, we insert a small padding to ensure that newly
+constructed AABBs always have a non-zero volume:
 
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
     ...
     class aabb {
       public:
         ...
-        aabb(const aabb& box0, const aabb& box1) {
-            x = interval(box0.x, box1.x);
-            y = interval(box0.y, box1.y);
-            z = interval(box0.z, box1.z);
+        aabb(const interval& _x, const interval& _y, const interval& _z)
+          : x(_x), y(_y), z(_z)
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++ highlight
+        {
+            pad_to_minimums();
         }
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
 
 
+        aabb(const point3& a, const point3& b) {
+            // Treat the two points a and b as extrema for the bounding box, so we don't require a
+            // particular minimum/maximum coordinate order.
+            x = interval(fmin(a[0],b[0]), fmax(a[0],b[0]));
+            y = interval(fmin(a[1],b[1]), fmax(a[1],b[1]));
+            z = interval(fmin(a[2],b[2]), fmax(a[2],b[2]));
+
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++ highlight
-        aabb pad() {
-            // Return an AABB that has no side narrower than some delta, padding if necessary.
-            double delta = 0.0001;
-            interval new_x = (x.size() >= delta) ? x : x.expand(delta);
-            interval new_y = (y.size() >= delta) ? y : y.expand(delta);
-            interval new_z = (z.size() >= delta) ? z : z.expand(delta);
+            pad_to_minimums();
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
+        }
+        ...
```

Comment:
I intentionally left the call out of that constructor, since I think we can assume the input `aabb` parameters already have a non-zero volume and so should the result of combining their intervals. I am happy to revisit that decision though and I could have overlooked a corner case.

Reply:
Agreed with @armansito.

Reply:
There's nothing that guarantees that the input intervals are non-singular, nor that a ≠ b. Agreed with @wlbksy.

Reply:
> Ugh. I'm batting 0 for 1000 today. Yes, since the other constructors ensure that all `aabb`s have non-zero dimensions, we are safe when composing such `aabb` objects.
> 
> I agree with @armansito's original reasoning and change.

Yeah. Original is sound. But since we two all gets confused here at first glimpse, maybe adding the `pad_to_minimums()` call here is not too overkill, and ease reader's mind.